python_examples/rebound.py

```diff
@@ -127,4 +149,158 @@ def step():
 def integrate(tmax):
     libias15.integrate(c_double(tmax))
 
+TWOPI = 2.*numpy.pi
+def mod2pi(f):
+    while f<0.:
+        f += TWOPI
+    while f>TWOPI:
+        f -= TWOPI
+    return f
+
+def get_E(e,M):
+    E = M if e<0.8 else numpy.pi
+    
+    F = E - e*numpy.sin(M) - M
+    for i in range(100):
+        E = E - F/(1.0-e*numpy.cos(E))
+        F = E - e*numpy.sin(E) - M
+        if numpy.fabs(F)<1e-16:
+            break
+    E = mod2pi(E)
+    return E
+
+def init_planet(star,       # central object (rebound.Particle object)
+                a,          # semimajor axis
+                e=None,     # eccentricity
+                omega=None, # argument of pericenter
+                f=None,     # true anomaly
+                inc=None,   # inclination
+                Omega=None, # longitude of ascending node
+                mass=None,  # planet mass
+                M=None):    # mean anomaly
+    '''Returns a particle structure to add into rebound for a given set of
+        orbital elements. 'star' and 'a' are required, others will default to zero
+        (with a warning). Only f OR M should be passed to specify the location of
+        the particle on its orbit.'''
+    
+    # first check for valid inputs and give warnings on defaults
+    if e == None: e = 0.; print('e not passed, setting to 0... (circular)')
+    if omega == None:
+        omega = 0. # only warn if orbit is non-circular
+        if e != 0.: print('omega not passed, setting to 0... (pericenter at ascending node)')
+    if inc == None: inc = 0.; print('inc not passed, setting to 0... (in xy plane)')
+    if Omega == None:
+        Omega = 0. # only warn if orbit is inclined
+        if inc != 0.: print('Omega not passed, setting to 0... (ascending node along x axis)')
+    if f != None and M != None: raise ValueError('Can only pass EITHER the true (f) or the mean (M) anomaly')
+    if f == None and M == None:
+        f = 0.;
+        if e != 0.:
+            print('Neither f nor M passed, setting initial angle from pericenter to 0...')
+        else:
+            print('Neither f nor M passed, setting particle on x axis...')
+    if mass == None: mass = 0.; print('mass not passed, setting to 0...')
+    
+    if not(0. <= e < 1.): raise ValueError('e must be in range [0,1)') # not sure if these equations work for parabolic/hyperbolic obits
+    if not(0. <= inc <= numpy.pi): raise ValueError('inc must be in range [0,pi]')
+    
+    if f == None: # need to calculate f
+        E = get_E(e,M)
+        f = mod2pi(2.*numpy.arctan(numpy.sqrt((1. + e)/(1. - e))*numpy.tan(0.5*E)))
+    
+    r = a*(1-e**2)/(1 + e*numpy.cos(f))
+    
+    # Murray & Dermott Eq. 2.122
+    _x  = star.x + r*(numpy.cos(Omega)*numpy.cos(omega+f) - numpy.sin(Omega)*numpy.sin(omega+f)*numpy.cos(inc))
```

Comment:
This ok, but inefficient. It might be worth caching all the `cos` and `sin` functions rather than calculating them multiple times. They can take a long time to compute.